### PR TITLE
Hide 'Add to Reusable Blocks' button when 'core/block' is disabled

### LINKED ISSUE #5893: Shared/Reusable Blocks do not show when using the `allowed_block_types` filter
If you limit the blocks which can be shown using the `allowed_block_types` filter then shared blocks do not appear in the 'Shared' block inserter tab.

### PR DESCRIPTION
**Pull request by noisysocks**
Fixes #5893.

Hides the _Add to Reusable Block_ button when `'core/block'` is not in the array returned by the `allowed_block_types` filter.

**To test:**

1. Add this to the bottom of `gutenberg.php`:

```php
add_filter( 'allowed_block_types', function() {
	return [ 'core/paragraph', 'core/image', 'core/quote' ];
} );
```

2. Check that you are not able to create or insert Reusable Blocks.

3. Now, change the above PHP code to look like this:

```php
add_filter( 'allowed_block_types', function() {
	return [ 'core/paragraph', 'core/image', 'core/quote', 'core/block' ];
} );
```

4. You should be able to once again create and insert Reusable Blocks.

5. Check that the _Add to Reusable Block_ does not appear when you select _More options_ on a _Classic_ block.

### CODE REVIEW

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (brandonpayton) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: It would be clearer to keep `isVisible` as a single expression because you can read one statement and understand completely what it means to be visible. Each subexpression can be commented if desired. Here, that knowledge is broken over three different control statements, so we have an increase in control flow and have to read multiple statements to follow what is happening and what it means to be visible.

In addition, each `if` statement encodes knowledge of what the return value looks like. If that shape changes, each return statement may need to be updated. If you'd like to keep the definition of `isVisible` broken up, perhaps it would be better assign `isVisible = false` in those cases and keep a single return statement.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (brandonpayton) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: Just for my own understanding:
Do you know how it is possible for us to have a `clientId` that doesn't have a corresponding block?

**Review by brandonpayton — APPROVED** [review]
I tested, and this works for me.

Please check out my review comments before merging. I argue for some changes to the code but am approving so you are able to merge the next time you look at this.

Thank you!

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: I've changed it to be a single expression again! 👍

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: I played around with this a little and this check seems to prevent an unknown property exception from being thrown when a static block is converted into a reusable one. 

I added a comment in 4bab114 which explains that.

Still, I'm not clear on how this happens. `withSelect` must be getting called after the static block is removed but before the block list has a chance to update... 🤔

**Comment by noisysocks** [comment]
Thanks @brandonpayton! I'm plopping this in despite some mystery surrounding the `every()` check (see https://github.com/WordPress/gutenberg/pull/10493#discussion_r225364191) since that check already exists in `master`.

**Comment by brandonpayton** [comment]
> Thanks @brandonpayton! I'm plopping this in despite some mystery surrounding the `every()` check (see [#10493 (comment)](https://github.com/WordPress/gutenberg/pull/10493#discussion_r225364191)) since that check already exists in `master`.

Thanks for taking the time to explain some of the mystery and for adding that comment.

**Pull request merged**
